Build error Ubuntu 14.04

----------------------------------------------------------------------------------------------------

| Required Info                         |                                                                |
|---------------------------------|------------------------------------------- |
| Camera Model                       |  D435  | 
| Firmware Version                   | (Open RealSense Viewer --> Click info) | 
| Operating System & Version |  Ubuntu 14 (32Bit)  | 
| Kernel Version (Linux Only)    |  4.4.0-141-generic  | 
| Platform                                 | PC  |
| SDK Version                            |  2.19.1                 | 
| Language                            |  C                          | 
| Segment			|  Robot                          | 

### Issue Description
Build fails in TM2 with compile error. 

```
/home/seville/librealsense/src/tm2/tm-device.cpp:627:69: error: no matching function for call to ‘copy_array(float [5], const float_t [5])’
         librealsense::copy_array(result.coeffs, tm_intrinsics.coeffs);
```

I followed the instructions from [here](https://github.com/IntelRealSense/librealsense/blob/master/doc/installation.md).

Please see attached file for complete log.
[librealsense.log](https://github.com/IntelRealSense/librealsense/files/2949836/librealsense.log)

--- a/CMake/global_config.cmake
+++ b/CMake/global_config.cmake
@@ -53,6 +53,10 @@ macro(global_set_flags)
         add_definitions(-DRS2_USE_CUDA)
     endif()
 
+    if (BUILD_INTERNAL_UNIT_TESTS)
+        add_definitions(-DBUILD_INTERNAL_UNIT_TESTS)
+    endif()
+
     if(FORCE_LIBUVC)
         set(BACKEND RS2_USE_LIBUVC_BACKEND)
         message(STATUS "Using libuvc (by force)")

--- a/common/tiny-profiler.h
+++ b/common/tiny-profiler.h
@@ -3,6 +3,7 @@
 #include <unordered_map>
 #include <chrono>
 #include <iostream>
+#include <cstring> // strlen
 
 class scoped_timer
 {

--- a/src/tm2/tm-device.cpp
+++ b/src/tm2/tm-device.cpp
@@ -822,7 +822,7 @@ namespace librealsense
         result.fx = tm_intrinsics.fx;
         result.fy = tm_intrinsics.fy;
         result.model = convertTm2CameraModel(tm_intrinsics.distortionModel);
-        librealsense::copy_array(result.coeffs, tm_intrinsics.coeffs);
+        librealsense::copy_array<true>(result.coeffs, tm_intrinsics.coeffs);
         return result;
     }
 
@@ -852,9 +852,9 @@ namespace librealsense
         {
             throw io_exception("Failed to read TM2 intrinsics");
         }
-        librealsense::copy_2darray(result.data, tm_intrinsics.data);
-        librealsense::copy_array(result.noise_variances, tm_intrinsics.noiseVariances);
-        librealsense::copy_array(result.bias_variances, tm_intrinsics.biasVariances);
+        librealsense::copy_2darray<true>(result.data, tm_intrinsics.data);
+        librealsense::copy_array<true>(result.noise_variances, tm_intrinsics.noiseVariances);
+        librealsense::copy_array<true>(result.bias_variances, tm_intrinsics.biasVariances);
 
         return result;
     }
@@ -895,8 +895,8 @@ namespace librealsense
             throw io_exception("Failed to read TM2 intrinsics");
         }
 
-        librealsense::copy_array(result.rotation, tm_extrinsics.rotation);
-        librealsense::copy_array(result.translation, tm_extrinsics.translation);
+        librealsense::copy_array<true>(result.rotation, tm_extrinsics.rotation);
+        librealsense::copy_array<true>(result.translation, tm_extrinsics.translation);
         reference_sensor_id = tm_extrinsics.referenceSensorId;
 
         return result;

--- a/src/types.h
+++ b/src/types.h
@@ -95,15 +95,38 @@ namespace librealsense
         operator std::string() const { return ss.str(); }
     };
 
-    template<typename T, size_t size>
-    inline size_t copy_array(T(&dst)[size], const T(&src)[size])
+    template<typename T, size_t sz>
+    constexpr size_t arr_size(T(&)[sz])
     {
-        assert(dst != nullptr && src != nullptr);
-        for (size_t i = 0; i < size; i++)
+        return sz;
+    }
+
+    template<typename T, size_t sz>
+    constexpr size_t arr_size_bytes(T(&)[sz])
+    {
+        return sz * sizeof(T);
+    }
+
+    template<typename T>
+    std::string array2str(T& data)
+    {
+        std::stringstream ss;
+        for (auto i = 0; i < arr_size(data); i++)
+            ss << " [" << i << "] = " << data[i] << "\t";
+        return ss.str();
+    }
+
+    template<typename T, typename S, size_t size_tgt, size_t size_src>
+    inline size_t copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
+    {
+        static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
+        static_assert(std::numeric_limits<S>::max() > std::numeric_limits<T>::max(), "copy_array cannot support narrowing conversions from T to S");
+
+        for (size_t i = 0; i < size_tgt; i++)
         {
-            dst[i] = src[i];
+            dst[i] = static_cast<T>(src[i]);
         }
-        return size;
+        return size_tgt;
     }
 
     template<typename T, size_t sizem, size_t sizen>
@@ -450,27 +473,6 @@ namespace librealsense
         unique_id(const unique_id&) = delete;
         unique_id& operator=(const unique_id&) = delete;
     };
-
-    template<typename T, int sz>
-    int arr_size(T(&)[sz])
-    {
-        return sz;
-    }
-
-    template<typename T, int sz>
-    int arr_size_bytes(T(&)[sz])
-    {
-        return sz*sizeof(T);
-    }
-
-    template<typename T>
-    std::string array2str(T& data)
-    {
-        std::stringstream ss;
-        for (auto i = 0; i < arr_size(data); i++)
-            ss << " [" << i << "] = " << data[i] << "\t";
-        return ss.str();
-    }
 
     typedef float float_4[4];
 

--- a/src/types.h
+++ b/src/types.h
@@ -95,16 +95,19 @@ namespace librealsense
         operator std::string() const { return ss.str(); }
     };
 
+    template<typename T>
+    constexpr size_t arr_size(T const&) { return 1; }
+
     template<typename T, size_t sz>
-    constexpr size_t arr_size(T(&)[sz])
+    constexpr size_t arr_size(T(&arr)[sz])
     {
-        return sz;
+        return sz * arr_size(arr[0]);
     }
 
     template<typename T, size_t sz>
-    constexpr size_t arr_size_bytes(T(&)[sz])
+    constexpr size_t arr_size_bytes(T(&arr)[sz])
     {
-        return sz * sizeof(T);
+        return arr_size(arr) * sizeof(T);
     }
 
     template<typename T>
@@ -116,12 +119,20 @@ namespace librealsense
         return ss.str();
     }
 
-    template<typename T, typename S, size_t size_tgt, size_t size_src>
-    inline size_t copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
-    {
-        static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
-        static_assert(std::numeric_limits<S>::max() > std::numeric_limits<T>::max(), "copy_array cannot support narrowing conversions from T to S");
+    // Auxillary function for compiler to highlight the invoking user code
+    template <typename T, typename S>
+    struct isNarrowing { static bool const value = std::numeric_limits<S>::max() > std::numeric_limits<T>::max(); };
 
+    // specialization for an array copy with a narrowing conversion
+    template<typename T, typename S, size_t size_tgt, size_t size_src>
+    inline typename std::enable_if<isNarrowing<T, S>::value, size_t > ::type copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
+    {
+#pragma message(": Note<librealsense::copy_array> using narrowing conversion")
+
+        static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
+        static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
+
+        assert(dst != nullptr && src != nullptr);
         for (size_t i = 0; i < size_tgt; i++)
         {
             dst[i] = static_cast<T>(src[i]);
@@ -129,18 +140,54 @@ namespace librealsense
         return size_tgt;
     }
 
-    template<typename T, size_t sizem, size_t sizen>
-    inline size_t copy_2darray(T(&dst)[sizem][sizen], const T(&src)[sizem][sizen])
+    template<typename T, typename S, size_t size_tgt, size_t size_src>
+    inline typename std::enable_if < !isNarrowing<T, S>::value, int> ::type copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
     {
+        static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
+        static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
+
         assert(dst != nullptr && src != nullptr);
-        for (size_t i = 0; i < sizem; i++)
+        for (size_t i = 0; i < size_tgt; i++)
         {
-            for (size_t j = 0; j < sizen; j++)
+            dst[i] = static_cast<T>(src[i]);
+        }
+        return size_tgt;
+    }
+
+    // specialization for an 2D-array copy with a narrowing conversion
+    template<typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
+    inline typename std::enable_if<isNarrowing<T, S>::value, size_t >::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
+    {
+#pragma message(": Note<librealsense::copy_2darray> using narrowing conversion")
+        static_assert((src_m && src_n && (tgt_n == src_n) && (tgt_m == src_m)), "copy_array requires similar non-zero size for target and source containers");
+        static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_2darray supports arithmetic types only");
+
+        assert(dst != nullptr && src != nullptr);
+        for (size_t i = 0; i < src_m; i++)
+        {
+            for (size_t j = 0; j < src_n; j++)
             {
                 dst[i][j] = src[i][j];
             }
         }
-        return sizem * sizen;
+        return src_m * src_n;
+    }
+
+    template<typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
+    inline typename std::enable_if<!isNarrowing<T, S>::value, int>::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
+    {
+        static_assert((src_m && src_n && (tgt_n == src_n) && (tgt_m == src_m)), "copy_array requires similar non-zero size for target and source containers");
+        static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
+
+        assert(dst != nullptr && src != nullptr);
+        for (size_t i = 0; i < src_m; i++)
+        {
+            for (size_t j = 0; j < src_n; j++)
+            {
+                dst[i][j] = src[i][j];
+            }
+        }
+        return src_m * src_n;
     }
 
     // Comparing parameter against a range of values of the same type
@@ -228,7 +275,7 @@ namespace librealsense
 #endif
     }
 
-#if RS2_BUILD_INTERNAL_UNIT_TESTS
+#ifdef BUILD_INTERNAL_UNIT_TESTS
 #define PRIVATE_TESTABLE public
 #else
 #define PRIVATE_TESTABLE private

--- a/src/types.h
+++ b/src/types.h
@@ -123,32 +123,15 @@ namespace librealsense
     template <typename T, typename S>
     struct isNarrowing { static bool constexpr value = std::numeric_limits<S>::max() > std::numeric_limits<T>::max(); };
 
-//     // specialization for an array copy with a narrowing conversion
-//     template<typename T, typename S, size_t size_tgt, size_t size_src>
-//     inline typename std::enable_if<isNarrowing<T, S>::value, size_t > ::type copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
-//     {
-// #ifdef _MSV_VER
-// #pragma message(": Note<librealsense::copy_array> using narrowing conversion")
-// #endif
-
-//         static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
-//         static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
-
-//         assert(dst != nullptr && src != nullptr);
-//         for (size_t i = 0; i < size_tgt; i++)
-//         {
-//             dst[i] = static_cast<T>(src[i]);
-//         }
-//         return size_tgt;
-//     }
-
-    template<typename T, typename S, size_t size_tgt, size_t size_src>
-    inline size_t copy_array(T(&dst)[size_tgt], const S(&src)[size_src], bool allow_narrowing = false)
+    template<const bool force_narrowing=false, typename T, typename S, size_t size_tgt, size_t size_src>
+    inline size_t copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
     {
         static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
         static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
-        if (isNarrowing<T, S>::value && !allow_narrowing)
-            static_assert(!isNarrowing<T, S>::value, "Passing narrowing conversion to copy_array requires setting the force flag on");
+        if (!force_narrowing && isNarrowing<T, S>::value)
+        {
+            static_assert(!(isNarrowing<T, S>::value && !force_narrowing), "Passing narrowing conversion to copy_array requires setting the force flag on");
+        }
 
         assert(dst != nullptr && src != nullptr);
         for (size_t i = 0; i < size_tgt; i++)
@@ -158,46 +141,26 @@ namespace librealsense
         return size_tgt;
     }
 
-    // specialization for an 2D-array copy with a narrowing conversion
-    template<typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
-    inline size_t copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n], bool allow_narrowing = false)
-    //inline typename std::enable_if<isNarrowing<T, S>::value, size_t >::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
+    template<const bool force_narrowing=false, typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
+    inline size_t copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
     {
-// #ifdef _MSV_VER
-// #pragma message(": Note<librealsense::copy_2darray> using narrowing conversion")
-// #endif
         static_assert((src_m && src_n && (tgt_n == src_n) && (tgt_m == src_m)), "copy_array requires similar non-zero size for target and source containers");
         static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_2darray supports arithmetic types only");
-        if (isNarrowing<T, S>::value && !allow_narrowing)
-            static_assert(!isNarrowing<T, S>::value, "Passing narrowing conversion to copy_2darray requires setting the force flag on");
+        if (isNarrowing<T, S>::value && !force_narrowing)
+        {
+            static_assert(!(isNarrowing<T, S>::value && !force_narrowing), "Passing narrowing conversion to copy_2darray requires setting the force flag on");
+        }
 
         assert(dst != nullptr && src != nullptr);
         for (size_t i = 0; i < src_m; i++)
         {
             for (size_t j = 0; j < src_n; j++)
             {
-                dst[i][j] = src[i][j];
+                dst[i][j] = static_cast<T>(src[i][j]);
             }
         }
         return src_m * src_n;
     }
-
-    // template<typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
-    // inline typename std::enable_if<!isNarrowing<T, S>::value, int>::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
-    // {
-    //     static_assert((src_m && src_n && (tgt_n == src_n) && (tgt_m == src_m)), "copy_array requires similar non-zero size for target and source containers");
-    //     static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
-
-    //     assert(dst != nullptr && src != nullptr);
-    //     for (size_t i = 0; i < src_m; i++)
-    //     {
-    //         for (size_t j = 0; j < src_n; j++)
-    //         {
-    //             dst[i][j] = src[i][j];
-    //         }
-    //     }
-    //     return src_m * src_n;
-    // }
 
     // Comparing parameter against a range of values of the same type
     // https://stackoverflow.com/questions/15181579/c-most-efficient-way-to-compare-a-variable-to-multiple-values

--- a/src/types.h
+++ b/src/types.h
@@ -121,30 +121,34 @@ namespace librealsense
 
     // Auxillary function for compiler to highlight the invoking user code
     template <typename T, typename S>
-    struct isNarrowing { static bool const value = std::numeric_limits<S>::max() > std::numeric_limits<T>::max(); };
+    struct isNarrowing { static bool constexpr value = std::numeric_limits<S>::max() > std::numeric_limits<T>::max(); };
 
-    // specialization for an array copy with a narrowing conversion
+//     // specialization for an array copy with a narrowing conversion
+//     template<typename T, typename S, size_t size_tgt, size_t size_src>
+//     inline typename std::enable_if<isNarrowing<T, S>::value, size_t > ::type copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
+//     {
+// #ifdef _MSV_VER
+// #pragma message(": Note<librealsense::copy_array> using narrowing conversion")
+// #endif
+
+//         static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
+//         static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
+
+//         assert(dst != nullptr && src != nullptr);
+//         for (size_t i = 0; i < size_tgt; i++)
+//         {
+//             dst[i] = static_cast<T>(src[i]);
+//         }
+//         return size_tgt;
+//     }
+
     template<typename T, typename S, size_t size_tgt, size_t size_src>
-    inline typename std::enable_if<isNarrowing<T, S>::value, size_t > ::type copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
+    inline size_t copy_array(T(&dst)[size_tgt], const S(&src)[size_src], bool allow_narrowing = false)
     {
-#pragma message(": Note<librealsense::copy_array> using narrowing conversion")
-
         static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
         static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
-
-        assert(dst != nullptr && src != nullptr);
-        for (size_t i = 0; i < size_tgt; i++)
-        {
-            dst[i] = static_cast<T>(src[i]);
-        }
-        return size_tgt;
-    }
-
-    template<typename T, typename S, size_t size_tgt, size_t size_src>
-    inline typename std::enable_if < !isNarrowing<T, S>::value, int> ::type copy_array(T(&dst)[size_tgt], const S(&src)[size_src])
-    {
-        static_assert((size_tgt && (size_tgt == size_src)), "copy_array requires similar non-zero size for target and source containers");
-        static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
+        if (isNarrowing<T, S>::value && !allow_narrowing)
+            static_assert(!isNarrowing<T, S>::value, "Passing narrowing conversion to copy_array requires setting the force flag on");
 
         assert(dst != nullptr && src != nullptr);
         for (size_t i = 0; i < size_tgt; i++)
@@ -156,11 +160,16 @@ namespace librealsense
 
     // specialization for an 2D-array copy with a narrowing conversion
     template<typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
-    inline typename std::enable_if<isNarrowing<T, S>::value, size_t >::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
+    inline size_t copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n], bool allow_narrowing = false)
+    //inline typename std::enable_if<isNarrowing<T, S>::value, size_t >::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
     {
-#pragma message(": Note<librealsense::copy_2darray> using narrowing conversion")
+// #ifdef _MSV_VER
+// #pragma message(": Note<librealsense::copy_2darray> using narrowing conversion")
+// #endif
         static_assert((src_m && src_n && (tgt_n == src_n) && (tgt_m == src_m)), "copy_array requires similar non-zero size for target and source containers");
         static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_2darray supports arithmetic types only");
+        if (isNarrowing<T, S>::value && !allow_narrowing)
+            static_assert(!isNarrowing<T, S>::value, "Passing narrowing conversion to copy_2darray requires setting the force flag on");
 
         assert(dst != nullptr && src != nullptr);
         for (size_t i = 0; i < src_m; i++)
@@ -173,22 +182,22 @@ namespace librealsense
         return src_m * src_n;
     }
 
-    template<typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
-    inline typename std::enable_if<!isNarrowing<T, S>::value, int>::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
-    {
-        static_assert((src_m && src_n && (tgt_n == src_n) && (tgt_m == src_m)), "copy_array requires similar non-zero size for target and source containers");
-        static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
+    // template<typename T, typename S, size_t tgt_m, size_t tgt_n, size_t src_m, size_t src_n>
+    // inline typename std::enable_if<!isNarrowing<T, S>::value, int>::type copy_2darray(T(&dst)[tgt_m][tgt_n], const S(&src)[src_m][src_n])
+    // {
+    //     static_assert((src_m && src_n && (tgt_n == src_n) && (tgt_m == src_m)), "copy_array requires similar non-zero size for target and source containers");
+    //     static_assert((std::is_arithmetic<S>::value) && (std::is_arithmetic<T>::value), "copy_array supports arithmetic types only");
 
-        assert(dst != nullptr && src != nullptr);
-        for (size_t i = 0; i < src_m; i++)
-        {
-            for (size_t j = 0; j < src_n; j++)
-            {
-                dst[i][j] = src[i][j];
-            }
-        }
-        return src_m * src_n;
-    }
+    //     assert(dst != nullptr && src != nullptr);
+    //     for (size_t i = 0; i < src_m; i++)
+    //     {
+    //         for (size_t j = 0; j < src_n; j++)
+    //         {
+    //             dst[i][j] = src[i][j];
+    //         }
+    //     }
+    //     return src_m * src_n;
+    // }
 
     // Comparing parameter against a range of values of the same type
     // https://stackoverflow.com/questions/15181579/c-most-efficient-way-to-compare-a-variable-to-multiple-values

--- a/unit-tests/internal/CMakeLists.txt
+++ b/unit-tests/internal/CMakeLists.txt
@@ -9,6 +9,7 @@ set (INTERNAL_TESTS_SOURCES
     internal-tests-main.cpp
     internal-tests-usb.cpp
     internal-tests-extrinsic.cpp
+	internal-tests-types.cpp
 )
 
 add_executable(${PROJECT_NAME} ${INTERNAL_TESTS_SOURCES})

--- a/unit-tests/internal/internal-tests-extrinsic.cpp
+++ b/unit-tests/internal/internal-tests-extrinsic.cpp
@@ -49,7 +49,7 @@ TEST_CASE("Extrinsic graph management", "[live]")
                     dev = list[j];
                 }
 
-                //std::cout << __LINE__ << " Iteration " << i << " : Extrinsic graph map size is " << b._streams.size() << std::endl;
+                std::cout << __LINE__ << " Iteration " << i << " : Extrinsic graph map size is " << b._streams.size() << std::endl;
                 for (auto&& snr : dev.query_sensors())
                 {
                     std::vector<rs2::stream_profile> profs;

--- a/unit-tests/internal/internal-tests-extrinsic.cpp
+++ b/unit-tests/internal/internal-tests-extrinsic.cpp
@@ -28,52 +28,67 @@ inline void require_identity_matrix(const float(&matrix)[9])
     for (int i = 0; i < 9; ++i) REQUIRE(matrix[i] == Approx(identity_matrix_3x3[i]));
 }
 
-TEST_CASE("Extrinsic graph management", "[live]")
+TEST_CASE("Extrinsic graph management", "[live][multicam]")
 {
     // Require at least one device to be plugged in
     rs2::context ctx;
     {
+        std::cout << "Extrinsic graph management started" << std::endl;
         auto list = ctx.query_devices();
         REQUIRE(list.size());
 
-        // For each device
+        std::map<std::string, size_t> extrinsic_graph_at_sensor;
         auto& b = environment::get_instance().get_extrinsics_graph();
-        for (int i = 0; i < 5000; i++)
-        {
-            //for (auto&& dev : list)
-            for (int j=0; j < list.size(); j++)
-            {
-                rs2::device dev{};
-                {
-                    //scoped_timer t1("Dev generation");
-                    dev = list[j];
-                }
+        auto init_size = b._streams.size();
+        std::cout << " Initial Extrinsic Graph size is " << init_size << std::endl;
 
-                std::cout << __LINE__ << " Iteration " << i << " : Extrinsic graph map size is " << b._streams.size() << std::endl;
+        for (int i = 0; i < 10; i++)
+        {
+            std::cout << "Iteration " << i << " : Extrinsic graph map size is " << b._streams.size() << std::endl;
+            // For each device
+            for (auto&& dev : list)
+            {
+                //std::cout << "Dev " << dev.get_info(RS2_CAMERA_INFO_NAME) << std::endl;
                 for (auto&& snr : dev.query_sensors())
                 {
                     std::vector<rs2::stream_profile> profs;
                     REQUIRE_NOTHROW(profs = snr.get_stream_profiles());
-                    //snr.get_info(RS2_CAMERA_INFO_NAME);
                     REQUIRE(profs.size() > 0);
 
-                    //std::cout << __LINE__ << " " << snr.get_info(RS2_CAMERA_INFO_NAME) <<" : Extrinsic graph map size is " << b._streams.size() << std::endl;
+                    std::string snr_id = snr.get_info(RS2_CAMERA_INFO_NAME);
+                    snr_id += snr.get_info(RS2_CAMERA_INFO_SERIAL_NUMBER);
+                    if (extrinsic_graph_at_sensor.count(snr_id))
+                    {
+                        CAPTURE(snr_id);
+                        CAPTURE(extrinsic_graph_at_sensor.at(snr_id));
+                        CAPTURE(b._streams.size());
+                        REQUIRE(b._streams.size() == extrinsic_graph_at_sensor.at(snr_id));
+                    }
+                    else
+                        extrinsic_graph_at_sensor[snr_id] = b._streams.size();
 
-                    //rs2_extrinsics extrin{};
-                    //try {
-                    //    auto prof = profs[0];
-                    //    extrin = prof.get_extrinsics_to(prof);
-                    //}
-                    //catch (const rs2::error &e) {
-                    //    // if device isn't calibrated, get_extrinsics must error out (according to old comment. Might not be true under new API)
-                    //    WARN(e.what());
-                    //    continue;
-                    //}
+//                    std::cout << __LINE__ << " " << snr.get_info(RS2_CAMERA_INFO_NAME) <<" : Extrinsic graph map size is " << b._streams.size() << std::endl;
 
-                    /*require_identity_matrix(extrin.rotation);
-                    require_zero_vector(extrin.translation);*/
+                    rs2_extrinsics extrin{};
+                    try {
+                        auto prof = profs[0];
+                        extrin = prof.get_extrinsics_to(prof);
+                    }
+                    catch (const rs2::error &e) {
+                        // if device isn't calibrated, get_extrinsics must error out (according to old comment. Might not be true under new API)
+                        WARN(e.what());
+                        continue;
+                    }
+
+                    require_identity_matrix(extrin.rotation);
+                    require_zero_vector(extrin.translation);
                 }
             }
         }
+
+        auto end_size = b._streams.size();
+        std::cout << " Final Extrinsic Graph size is " << end_size << std::endl;
+        //REQUIRE(end_size == init_size); TODO doesn't pass yet
+        WARN("TODO: Graph size shall be preserved: init " << init_size << " != final " << end_size);
     }
 }

--- a/unit-tests/internal/internal-tests-types.cpp
+++ b/unit-tests/internal/internal-tests-types.cpp
@@ -73,8 +73,6 @@ TEST_CASE("copy_2darray", "")
     float       tgt_float[2][6] = { 0 };
     double      tgt_double[2][6] = { 0 };
 
-    std::cout << " array size is " << arr_size(src_float) << std::endl;
-
     // Check that precision is preserved when using identical or a larger data type
     elem = librealsense::copy_2darray(tgt_float, src_float);
     REQUIRE(elem == src_size);

--- a/unit-tests/internal/internal-tests-types.cpp
+++ b/unit-tests/internal/internal-tests-types.cpp
@@ -16,7 +16,7 @@
 using namespace librealsense;
 using namespace librealsense::platform;
 
-TEST_CASE("copy_array", "")
+TEST_CASE("copy_array", "[code]")
 {
     size_t elem = 0;
     float src_float[]   = { 1.1f, 2.2f, 3.3f, -5.5f, 0.f, 123534.f };
@@ -45,7 +45,7 @@ TEST_CASE("copy_array", "")
 
     // Check that precision is lost when narrowing data type
     memset(&tgt_float, 0, sizeof(float)*src_size);
-    elem = librealsense::copy_array(tgt_float, src_double);
+    elem = librealsense::copy_array<true>(tgt_float, src_double);
     REQUIRE(elem == src_size);
 
     for (size_t i = 0; i < src_size; ++i)
@@ -56,7 +56,7 @@ TEST_CASE("copy_array", "")
     }
 }
 
-TEST_CASE("copy_2darray", "")
+TEST_CASE("copy_2darray", "[code]")
 {
     size_t elem = 0;
     float src_float[2][6] = { { 1.1f, 2.2f, 3.3f, -5.5f, 0.f, 123534.f },
@@ -91,7 +91,7 @@ TEST_CASE("copy_2darray", "")
 
     // Check that precision is lost when narrowing data type
     memset(&tgt_float, 0, sizeof(float)*src_size);
-    elem = librealsense::copy_2darray(tgt_float, src_double);
+    elem = librealsense::copy_2darray<true>(tgt_float, src_double);
     REQUIRE(elem == src_size);
 
     for (size_t i = 0; i < h; ++i)

--- a/unit-tests/internal/internal-tests-types.cpp
+++ b/unit-tests/internal/internal-tests-types.cpp
@@ -1,0 +1,106 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2015 Intel Corporation. All Rights Reserved.
+
+#include "catch/catch.hpp"
+#include <cmath>
+#include <iostream>
+#include <chrono>
+#include <ctime>
+#include <algorithm>
+#include <type_traits>
+#include <librealsense2/rs.hpp>
+#include <librealsense2/hpp/rs_sensor.hpp>
+#include "../../common/tiny-profiler.h"
+#include "./../src/environment.h"
+
+using namespace librealsense;
+using namespace librealsense::platform;
+
+TEST_CASE("copy_array", "")
+{
+    size_t elem = 0;
+    float src_float[]   = { 1.1f, 2.2f, 3.3f, -5.5f, 0.f, 123534.f };
+    double src_double[] = { 0.00000000000000000001, -2222222222222222222222222.222222, 334529087543.30875246784, -5.51234533524532345254645234256, 5888985940.4535, -0.0000000000001 };
+
+    constexpr size_t src_size = arr_size(src_float);
+
+    float       tgt_float[src_size]     = { 0 };
+    double      tgt_double[src_size]    = { 0 };
+
+    // Check that precision is preserved when using identical or a larger data type
+    elem = librealsense::copy_array(tgt_float, src_float);
+    REQUIRE(elem == src_size);
+    elem = librealsense::copy_array(tgt_double, src_float);
+    REQUIRE(elem == src_size);
+
+    // Check that precision is preserved when expanding o using similar data types
+    for (size_t i = 0; i < src_size; ++i)
+    {
+        CAPTURE(src_double[i]);
+        CAPTURE(tgt_float[i]);
+        CAPTURE(tgt_double[i]);
+        REQUIRE(src_float[i] == (tgt_float[i]));
+        REQUIRE(src_float[i] == (tgt_double[i]));
+    }
+
+    // Check that precision is lost when narrowing data type
+    memset(&tgt_float, 0, sizeof(float)*src_size);
+    elem = librealsense::copy_array(tgt_float, src_double);
+    REQUIRE(elem == src_size);
+
+    for (size_t i = 0; i < src_size; ++i)
+    {
+        CAPTURE(src_double[i]);
+        CAPTURE(tgt_float[i]);
+        REQUIRE(src_double[i] != tgt_float[i]);
+    }
+}
+
+TEST_CASE("copy_2darray", "")
+{
+    size_t elem = 0;
+    float src_float[2][6] = { { 1.1f, 2.2f, 3.3f, -5.5f, 0.f, 123534.f },
+                                { 764.07f, -2342.2f, 2453.653f, -445.215f, 0.11324f, 1224.8209f } };
+    double src_double[2][6] = { { -254354352341.1, 765732052.21124, 3.4432007654764633233554, -524432.5432650645, 0.0000000000000000112, 123534.4234254673 },
+                                { 764.07654343263, -242675463465342.243, -9876322453.6342453, -42315432545.2153542, 0.1345521543251324, -0.0000123413242329 } };
+
+    constexpr size_t src_size = arr_size(src_float);
+
+    // Get the internal dimension sizes
+    auto h = std::extent<decltype(src_float),0>::value;
+    auto w = std::extent<decltype(src_float),1>::value;
+    
+    float       tgt_float[2][6] = { 0 };
+    double      tgt_double[2][6] = { 0 };
+
+    std::cout << " array size is " << arr_size(src_float) << std::endl;
+
+    // Check that precision is preserved when using identical or a larger data type
+    elem = librealsense::copy_2darray(tgt_float, src_float);
+    REQUIRE(elem == src_size);
+    elem = librealsense::copy_2darray(tgt_double, src_float);
+    REQUIRE(elem == src_size);
+
+    for (size_t i = 0; i < h; ++i)
+        for (size_t j = 0; j < w; ++j)
+        {
+            CAPTURE(src_float[i][j]);
+            CAPTURE(tgt_float[i][j]);
+            CAPTURE(tgt_double[i][j]);
+            REQUIRE(src_float[i][j] == (tgt_float[i][j]));
+            REQUIRE(src_float[i][j] == (tgt_double[i][j]));
+        }
+
+    // Check that precision is lost when narrowing data type
+    memset(&tgt_float, 0, sizeof(float)*src_size);
+    elem = librealsense::copy_2darray(tgt_float, src_double);
+    REQUIRE(elem == src_size);
+
+    for (size_t i = 0; i < h; ++i)
+        for (size_t j = 0; j < w; ++j)
+        {
+            CAPTURE(src_double[i][j]);
+            CAPTURE(tgt_float[i][j]);
+            REQUIRE(src_double[i][j] != tgt_float[i][j]);
+        }
+}

